service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/cmo/UploadCMOController.java

```diff
@@ -0,0 +1,147 @@
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        data.putAll(cmoService.getInitialPageData(caseData.getPastHearings(), caseData.getDraftUploadedCMOs()));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(data)
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleLabelMidEvent(@RequestBody CallbackRequest request) {
+        Map<String, Object> data = request.getCaseDetails().getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        // update judge and hearing labels
+        Object dynamicList = caseData.getPastHearingList();
+        List<Element<HearingBooking>> hearings = cmoService.getHearingsWithoutCMO(caseData.getPastHearings(),
+            caseData.getDraftUploadedCMOs());
+        UUID selectedHearing = cmoService.getSelectedHearingId(dynamicList, hearings);
+        data.putAll(cmoService.getJudgeAndHearingDetails(selectedHearing, hearings));
+
+        if (!(dynamicList instanceof DynamicList)) {
+            // reconstruct dynamic list
+            data.put("pastHearingList", cmoService.buildDynamicList(hearings, selectedHearing));
+        }
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(data)
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody CallbackRequest request) {
+        CaseDetails caseDetails = request.getCaseDetails();
+        Map<String, Object> data = caseDetails.getData();
+        CaseData caseData = mapper.convertValue(data, CaseData.class);
+        List<Element<CaseManagementOrder>> draftCMOs = caseData.getDraftUploadedCMOs();
+        List<Element<HearingBooking>> hearingsWithoutCMO = cmoService.getHearingsWithoutCMO(
+            caseData.getPastHearings(), draftCMOs
+        );
+
+        if (!hearingsWithoutCMO.isEmpty()) {
+            List<Element<HearingBooking>> hearings = caseData.getHearingDetails();
+            UUID selectedHearingId = cmoService.getSelectedHearingId(caseData.getPastHearingList(),
+                hearingsWithoutCMO);
+            HearingBooking hearing = cmoService.getSelectedHearing(selectedHearingId, hearingsWithoutCMO);
+
+            Element<CaseManagementOrder> element = element(from(caseData.getUploadedCaseManagementOrder(),
+                hearing, time.now().toLocalDate()));
+
+            UUID uuid = cmoService.mapToHearing(selectedHearingId, hearings, element);
+
+            if (uuid != null) {
+                // overwrite old draft CMO
+                int index = -1;
+                for (int i = 0; i < draftCMOs.size(); i++) {
+                    if (draftCMOs.get(i).getId().equals(uuid)) {
+                        index = i;
+                        break;
+                    }
+                }
```

Comment:
I would argue it's too much logic here.

The reason is why we want to decouple logic from controller is - in modern applications, you can have often multiple entry points to the system (i.e. REST API / Messaging system as best example), it's best if you can easily invoke the same processing from two different places.

I know we want be there here, but it's always to have the good principles. And possibly you can avoid some controller tests by pushing this further down, as this should be tested by service tests.

Reply:
Yeah, forgot to move this into the service, thanks for pointing it out